robot_smach_states/src/robot_smach_states/human_interaction/human_interaction.py

```diff
@@ -65,21 +82,75 @@ def execute(self, userdata=None):
             rospy.logerr("sentence = None, not saying anything...")
             return "spoken"
 
-        if not isinstance(self.sentence, str) and isinstance(self.sentence, list):
-            self.sentence = random.choice(self.sentence)
+        if hasattr(self.sentence, "resolve"):
+            sentence = self.sentence.resolve()
+            self._check_place_holders(sentence)
+        else:
+            sentence = self.sentence
+
+        if not isinstance(sentence, str) and isinstance(sentence, list):
+            sentence = random.choice(sentence)
 
-        sentence = str(self.sentence.resolve() if hasattr(self.sentence, "resolve") else self.sentence)
+        resolved_ph = {k: v.resolve() for k, v in self.ph_designators.items()}
+        sentence = sentence.format(**resolved_ph)
 
         if self.look_at_standing_person:
             self.robot.head.look_at_standing_person()
-        self.robot.speech.speak(sentence, self.language, self.personality, self.voice, self.mood, self.block)
-
-        # robot.head.cancel_goal()
-        # ToDo: hack
-        rospy.sleep(len(sentence)*0.05)
+        self.robot.speech.speak(str(sentence), self.language, self.personality, self.voice, self.mood, self.block)
 
         return "spoken"
 
+    def _check_place_holders(self, sentence):
+        if isinstance(sentence, list):
+            for sen in sentence:
+                self._check_place_holders(sen)
+            return
+
+        place_holders = set(x[1] for x in Formatter().parse(sentence) if x[1] is not None)
+        missing_ph = place_holders - set(self.ph_designators.keys())
+        if missing_ph:
+            raise RuntimeError("Not all named place holders are provided, missing: {}".
+                               format(", ".join(map(str, missing_ph))))
+
+
+class Say(SayFormatted):
+    """
+    Say a sentence or pick a random one from a list.
+
+    >>> from mock import MagicMock
+    >>> robot = MagicMock()
+    >>> robot.speech = MagicMock()
+    >>> robot.speech.speak = MagicMock()
+    >>>
+    >>> say = Say(robot, ["a", "b", "c"])
+    >>> #Repeat command 50 times, every time it should succeed and return "spoken"
+    >>> outcomes = [say.execute() for i in range(50)]
+    >>> assert all(outcome == "spoken" for outcome in outcomes)
+    >>>
+    >>> say2 = Say(robot, ds.VariableDesignator('aap'))
+    >>> say2.execute()
+    'spoken'
+    >>> robot.speech.speak.assert_called_with('aap', None, None, None, None, True)
```

Comment:
The assert_called_with does not exist as far as I know

Reply:
This is part of MagicMock